### PR TITLE
feat: add proxy metadata required rule for Hermeto-fetched packages

### PR DESCRIPTION
**Pull request by joejstuart**
## Summary

- Export `component_found_by_hermeto`, `package_found_by_hermeto`, and `is_registry_dependency` as public helpers in `policy/lib/sbom/sbom.rego`, refactoring `rpm.rego` to delegate to them
- Add `proxy_metadata_required` deny rule for CycloneDX SBOMs: denies when a qualifying package (found by Hermeto, proxy-enabled PURL type, registry dependency) has no `externalReference` of type `"distribution"`
- Add `proxy_metadata_required` deny rule for SPDX SBOMs: denies when a qualifying package has empty/missing `sourceInfo` field
- Update EC-1725 `allowed_proxy_urls` rules in both CycloneDX and SPDX to add the same three qualifying filters (found by Hermeto, proxy-enabled PURL type, registry dependency)

Resolves: EC-1723

## Test plan

- [x] All 841 policy tests pass with 100% coverage
- [x] `proxy_metadata_required` denies when qualifying package has no proxy metadata (CycloneDX and SPDX)
- [x] Rule passes for packages not found by Hermeto
- [x] Rule passes for PURL types not in `proxy_enabled_purl_types`
- [x] Rule passes for direct downloads (`download_url` qualifier) or git dependencies (`vcs_url` qualifier)
- [x] `allowed_proxy_urls` rules updated with same qualifying filters
- [x] Documentation regenerated

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: policy/release/sbom_spdx/sbom_spdx.rego [hunk 1]
```diff
@@ -304,6 +309,52 @@ deny contains result if {
...
+
+	source_info := object.get(pkg, "sourceInfo", "")
+	source_info == ""
+
+	result := metadata.result_helper_with_term(
+		rego.metadata.chain(),
+		[purl],
+		purl,
+	)
+}
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/release/sbom_spdx/sbom_spdx.rego:351]: The spec says when it's empty. Nothing else.

File: policy/release/sbom_spdx/sbom_spdx.rego [hunk 2]
```diff
@@ -304,6 +309,52 @@ deny contains result if {
...
+	externalref.referenceType == "purl"
+
+	purl := externalref.referenceLocator
+	parsed_purl := ec.purl.parse(purl)
+	parsed_purl.type in proxy_enabled
+
+	sbom.is_registry_dependency(parsed_purl)
+
+	source_info := object.get(pkg, "sourceInfo", "")
+	source_info == ""
```
Comment (st3penta) [inline policy/release/sbom_spdx/sbom_spdx.rego:344]: A dependency with `sourceInfo: "x"` and `downloadLocation: "NOASSERTION"` would skip proxy enforcement, because it passes both proxy rules:
- `proxy_metadata_required` passes because sourceInfo is non-empty
- `allowed_proxy_urls` skips NOASSERTION

Consider validating `sourceInfo` content (e.g. minimum length or URL-like pattern), or reconsidering the NOASSERTION skip for deps that have `sourceInfo` set.

File: policy/release/sbom_cyclonedx/sbom_cyclonedx.rego [hunk 3]
```diff
@@ -318,6 +322,54 @@ deny contains result if {
...
+		rego.metadata.chain(),
+		[purl],
+		purl,
+	)
+}
+
+_has_distribution_reference(component) if {
+	some reference in component.externalReferences
+	reference.type == "distribution"
+}
```
Comment (st3penta) [inline policy/release/sbom_cyclonedx/sbom_cyclonedx.rego:368]: This checks only for `type == "distribution"` without filtering NOASSERTION URLs.
Similarly to the other comment, a component with a NOASSERTION distribution skips all proxy enforcement.

```suggestion
_has_distribution_reference(component) if {
	some reference in component.externalReferences
	reference.type == "distribution"
	object.get(reference, "url", "") != "NOASSERTION"
}
```

File: policy/lib/sbom/sbom.rego [hunk 4]
```diff
@@ -395,6 +395,42 @@ rule_data_errors contains error if {
...
+# hermeto_found_by_property generates the CycloneDX property object used
+# to identify components fetched by the given tool name.
+hermeto_found_by_property(name) := _hermeto_found_by_property(name)
+
+_hermeto_found_by_property(name) := {
+	"name": sprintf("%s:found_by", [name]),
+	"value": name,
+}
+
+_hermeto_names := ["cachi2", "hermeto"]
```
Comment (st3penta) [inline policy/lib/sbom/sbom.rego:434]: nitpick: `allowed_package_sources` in both `sbom_cyclonedx.rego` and `sbom_spdx.rego` still uses inline hermeto detection instead of these new centralized helpers. If `_hermeto_names` is updated, those rules would be missed. 
- https://github.com/joejstuart/ec-policies/blob/57c3c29c5903c69f82437b805f511a9708aed656/policy/release/sbom_cyclonedx/sbom_cyclonedx.rego#L252-L253
- https://github.com/joejstuart/ec-policies/blob/57c3c29c5903c69f82437b805f511a9708aed656/policy/release/sbom_spdx/sbom_spdx.rego#L195-L196

File: policy/release/sbom_cyclonedx/sbom_cyclonedx_test.rego [hunk 5]
```diff
@@ -660,167 +660,306 @@ _sbom_1_6_attestation := json.patch(_sbom_1_5_attestation, [
...
+		"code": "sbom_cyclonedx.proxy_metadata_required",
+		"term": "pkg:maven/org.example/lib@1.0",
+		# regal ignore:line-length
+		"msg": `Package pkg:maven/org.example/lib@1.0 is missing proxy metadata (no externalReference of type "distribution")`,
+	}}
+
+	att := json.patch(_sbom_1_5_attestation, [{
+		"op": "add",
+		"path": "/statement/predicate/components/-",
+		"value": _cdx_hermeto_component("pkg:maven/org.example/lib@1.0", []),
```
Comment (st3penta) [inline policy/release/sbom_cyclonedx/sbom_cyclonedx_test.rego:873]: Consider adding a test where references exist but none have `type == "distribution"`, like `[{"type": "vcs", "url": "https://github.com/..."}]`

File: policy/release/sbom_spdx/sbom_spdx_test.rego [hunk 6]
```diff
@@ -541,3 +599,121 @@ _spdx_proxy_rule_data := {
...
+		"code": "sbom_spdx.proxy_metadata_required",
+		"term": "pkg:maven/org.example/lib@1.0",
+		# regal ignore:line-length
+		"msg": `Package pkg:maven/org.example/lib@1.0 is missing proxy metadata (sourceInfo is empty or missing)`,
+	}}
+
+	att := json.patch(_sbom_attestation, [{
+		"op": "add",
+		"path": "/statement/predicate/packages/-",
+		"value": _spdx_hermeto_package("pkg:maven/org.example/lib@1.0", ""),
```
Comment (st3penta) [inline policy/release/sbom_spdx/sbom_spdx_test.rego:616]: Consider adding a test where `sourceInfo` is absent from the package object entirely

**Review by st3penta — COMMENTED** [review]
Overall pretty good, but there are a couple of possible gaps in the proxy enforcement. I also left a nitpick and a couple of test case suggestions

File: policy/lib/sbom/rpm.rego [hunk 7]
```diff
@@ -44,29 +44,11 @@ _is_rpmish(purl) if {
...
 }
 
-# CycloneDX style
-_component_found_by_cachi2(component) if {
-	some property in component.properties
-	some cachi2_name in _cachi2_names
-	property == _cachi2_found_by_property(cachi2_name)
-} else := false
+# CycloneDX style - delegates to the public helper in sbom.rego
+_component_found_by_cachi2(component) := component_found_by_hermeto(component)
```
Comment (simonbaird) [inline policy/lib/sbom/rpm.rego]: Since we're changing this code, maybe it's a good time to retire the old cachi2 name:


```suggestion
_component_found_by_hermeto(component) := component_found_by_hermeto(component)
```

File: policy/lib/sbom/sbom.rego (hunk 4 above)
Comment (simonbaird) [inline policy/lib/sbom/sbom.rego:434]: For people who don't know the history, let's keep a comment:

```suggestion
# Hermeto used to be known as cachi2, so we recognize the
# old name for backwards compatibility
_hermeto_names := ["cachi2", "hermeto"]
```

File: policy/release/sbom_spdx/sbom_spdx.rego (hunk 2 above)
Comment (joejstuart) [reply policy/release/sbom_spdx/sbom_spdx.rego:344]: The `proxy_metadata_required` rule checks that `sourceInfo` is non-empty, which is what the spec defines.

File: policy/release/sbom_cyclonedx/sbom_cyclonedx.rego (hunk 3 above)
Comment (joejstuart) [reply policy/release/sbom_cyclonedx/sbom_cyclonedx.rego:368]: Done, added the NOASSERTION filter.

File: policy/lib/sbom/sbom.rego (hunk 4 above)
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:434]: Good catch, switched both rules to use the centralized helpers.

File: policy/release/sbom_cyclonedx/sbom_cyclonedx_test.rego (hunk 5 above)
Comment (joejstuart) [reply policy/release/sbom_cyclonedx/sbom_cyclonedx_test.rego:873]: Added `test_proxy_metadata_required_cdx_non_distribution_refs_denied`.

File: policy/release/sbom_spdx/sbom_spdx_test.rego (hunk 6 above)
Comment (joejstuart) [reply policy/release/sbom_spdx/sbom_spdx_test.rego:616]: Added `test_proxy_metadata_required_spdx_absent_source_info_denied`.

File: policy/lib/sbom/rpm.rego (hunk 7 above)
Comment (joejstuart) [reply policy/lib/sbom/rpm.rego]: Went a step further and removed the wrappers entirely — now calling the public helpers directly.

File: policy/lib/sbom/sbom.rego (hunk 4 above)
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:434]: Done, added the comment.

**Review by simonbaird — APPROVED** [review]
Lgtm.
